### PR TITLE
feat: glob and directory support for `deno check` and `deno cache` cli arg paths

### LINKED ISSUE #20813: feat: support passing directories to `deno check`
I find myself often ending up in a situation where I want to check if all types in a project are valid. Instinctively, I usually type `deno check` and then I'm getting an error that I didn't pass arguments. In my mind I'm thinking "ah right, so I want to type check everything in the current directory". This leads me to type `deno check .`, but this throws another error because directories are not supported. But shoot, I don't have any file which imports every other file in the project. So I'm not sure how to type check the whole project with one command.

I found a workaround though: If you have an extensive test suite, there is a high chance that all files in the project are imported at some point. If that's the case then running `deno test` is a better mechanism to type check a project than `deno check`, which seems odd.

Feature request:
- Add support for passing directories to `deno check`
- If no arguments are provided, use the current directory `.` as the default argument.

### LINKED ISSUE #24668: globbing support when type checking
As far as I can tell, `deno check` expects a file as entry point; it does not support targeting directories (failing silently; i.e. this limitation isn't obvious). Globbing might be used instead, but appears to be unsupported - unlike with adjacent commands (cf. #19102), which seems like a confusing inconsistency.

Thus projects without an obvious entry point covering all code (e.g. libraries with independent modules) need an out-of-band solution for module discovery, often resulting in shell scripts with limited portability.

Ideally, `check` should work just like `lint`, `fmt` and `test`:

```shell
$ deno lint "./src/**/*.ts"
$ deno check "./src/**/*.ts"
```

Perhaps it should also support targeting directories, though that might be a separate issue.

(For additional context, I've started documenting the way I've come to appreciate [Deno as general-purpose tooling](https://prepitaph.org/wip/deno-testing/) even beyond Deno-as-runtime projects, notably for client-side code.)

Version: Deno 1.45.2

### PR DESCRIPTION
**Pull request by yazan-abdalrahman**
Closes #24668
Closes #20813

### CODE REVIEW

**Comment by yazan-abdalrahman** [comment]
@marvinhagemeister please review this PR?

File: cli/graph_container.rs [hunk 1]
```diff
@@ -94,6 +98,51 @@ impl MainModuleGraphContainer {
...
   }
 
+  fn resolve_files(
+    &self,
+    patterns: &[String],
+  ) -> Result<Vec<String>, AnyError> {
+    let mut resolved_files = Vec::new();
+
+    for pattern in patterns {
+      let expanded = glob::glob(pattern).expect("Failed to read glob pattern");
```
Comment (dsherret) [inline cli/graph_container.rs:108]: Ideally this code should do something similar to what lint and fmt does (so not using glob directly and instead use `PathOrPatternSet`).

Comment (yazan-abdalrahman) [reply cli/graph_container.rs:108]: ok

Comment (yazan-abdalrahman) [reply cli/graph_container.rs]: done; please review it again.

**Comment by yazan-abdalrahman** [comment]
@dsherret 
can we merge it?

File: cli/graph_container.rs [hunk 2]
```diff
@@ -155,3 +159,45 @@ impl<'a> ModuleGraphUpdatePermit for MainModuleGraphUpdatePermit<'a> {
...
+        let base_path = pattern.base_path();
+        visit_dirs(&base_path, &pattern_set, &mut result_files)?;
+      }
+      PathOrPattern::NegatedPath(_) => {}
+      PathOrPattern::RemoteUrl(url) => result_files.push(url.to_string()),
+    }
+  }
+
+  Ok(result_files)
+}
```
Comment (dsherret) [inline cli/graph_container.rs:203]: We don't need this method. We can use `FileCollector` and pass it the paths and patterns. Right now this code does things like visiting the vendor or node_modules directory. See the lint or fmt subcommands for an example of using `FileCollector`

**Review by K1ngKP — APPROVED** [review]
This issue is solved?

**Review by K1ngKP — APPROVED** [review]

**Comment by yazan-abdalrahman** [comment]
@dsherret done with FileCollector

image

**Comment by yazan-abdalrahman** [comment]
@dsherret 

I used the FileCollector with the same lint and fmt, but it did not support the requirement

`$ deno lint "./src/**/*.ts"
$ deno check "./src/**/*.ts"`  
 it should type check all files that match the glob pattern but in lint and format it didn't do that  

check this
image

image

image

  
I think the previous method is better than using FileCollector with the same as format and lint.

**Comment by marvinhagemeister** [comment]
@yazan-abdalrahman Could it be because the pattern is different in the screenshots? It's missing a `*` in the glob.

```diff
- deno check ".\src\*\*.ts"
+ deno check ".\src\**\*.ts"
```

**Comment by yazan-abdalrahman** [comment]
> ```diff
> ".\src\**\*.ts"
> ```

@marvinhagemeister 
image
same lint

**Review by dsherret — APPROVED** [review]
LGTM

**Pull request merged**